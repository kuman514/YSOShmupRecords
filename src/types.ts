export interface NavNodeInfo {
  id: string;
  label: string;
  childNavNodeIds: NavNodeInfo['id'][];
}

export interface ShmupRecord {
  id: string;
  when: Date;
  subjectId: string;
  stage: string;
  /**
   * Why score should be string?
   * In case of exceeding `Number.MAX_SAFE_INTEGER`, which equals to `9007199254740991`.
   */
  score: string;
  byWhat: string;
  comment: string;
  thumbnailUrl: string;
  originalImageUrl: string;
  tweetUrl: string;
  youtubeUrl?: string;
  specialTags?: string[];
}
