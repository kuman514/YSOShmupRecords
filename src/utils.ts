import { monthTitle } from '^/constants/texts';

export function convertDateToString(date: Date) {
  return `${date.getFullYear()}년 ${
    monthTitle[date.getMonth()]
  } ${date.getDate()}일`;
}
