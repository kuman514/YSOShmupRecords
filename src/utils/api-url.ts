/**
 * @todo
 * Remove hardcoded api url
 */
export function getAPIURL(...paths: string[]) {
  return `https://read-only-api-endpoints-kuman514.vercel.app/yso-shmup-records${paths.join(
    '/'
  )}`;
}
