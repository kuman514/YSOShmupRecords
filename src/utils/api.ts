export function getAPIURL(...paths: string[]) {
  return `${import.meta.env.VITE_API_URL}/${paths.join('/')}`;
}
