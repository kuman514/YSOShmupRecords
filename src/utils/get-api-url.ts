export function getAPIURL(...paths: string[]) {
  return paths.join('/');
}
