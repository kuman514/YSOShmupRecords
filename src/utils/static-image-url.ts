export function getStaticImageUrl(...paths: string[]) {
  return `${import.meta.env.VITE_STATIC_IMAGE_URL}/${paths.join('/')}`;
}
