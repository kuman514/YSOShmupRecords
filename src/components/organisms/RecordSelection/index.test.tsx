import React from 'react';
import { cleanup, fireEvent, render, screen } from '@testing-library/react';
import { createMemoryRouter, RouterProvider } from 'react-router-dom';
import { describe, it, beforeEach, expect } from 'vitest';
import 'jest-styled-components';

import { ShmupRecordPreview } from '^/types';
import { RecordSelection } from '.';
import { convertDateToString } from '^/utils/date-to-string';

const recordIdsForTest: string[] = ['2023-10-28', '2023-05-14', '2022-11-21'];

const recordPreviewForTest = recordIdsForTest.map(
  (recordId): ShmupRecordPreview => ({
    id: recordId,
    title: convertDateToString(new Date(recordId)),
    imageUrl: `${recordId}.jpg`,
  })
);

describe('RecordSelection', () => {
  beforeEach(() => {
    cleanup();
  });

  it('should match snapshot with empty recordIds', () => {
    const routes = [
      {
        path: '/',
        element: <RecordSelection recordPreviews={[]} />,
      },
    ];

    const router = createMemoryRouter(routes, {
      initialEntries: ['/'],
      initialIndex: 0,
    });

    const { container } = render(<RouterProvider router={router} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should match snapshot with non-empty recordIds', () => {
    const routes = [
      {
        path: '/',
        element: <RecordSelection recordPreviews={recordPreviewForTest} />,
      },
    ];

    const router = createMemoryRouter(routes, {
      initialEntries: ['/'],
      initialIndex: 0,
    });

    const { container } = render(<RouterProvider router={router} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should route on click record selection item', async () => {
    const routes = [
      {
        path: '/',
        element: <RecordSelection recordPreviews={recordPreviewForTest} />,
      },
      {
        path: '/2023-05-14',
        element: <div>Correct!</div>,
      },
    ];

    const router = createMemoryRouter(routes, {
      initialEntries: ['/', '/2023-05-14'],
      initialIndex: 0,
    });

    render(<RouterProvider router={router} />);
    fireEvent.click(await screen.findByText(/2023년 5월 14일/i));
    expect(await screen.findByText(/Correct!/i)).not.toBeNull();
  });
});
