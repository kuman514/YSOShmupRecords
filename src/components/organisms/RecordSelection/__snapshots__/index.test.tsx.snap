// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`RecordSelection > should match snapshot with empty recordIds 1`] = `
.c4 {
  all: unset;
  box-sizing: border-box;
  font-size: 16px;
  font-weight: 500;
  padding: 8px 16px;
  color: var(--white-color);
  background-color: var(--primary-color);
  border: 1px solid var(--primary-color);
  border-radius: 4px;
  cursor: pointer;
}

.c4:hover {
  background-color: var(--hovering-color);
  color: var(--white-color);
}

.c4:disabled {
  border: none;
  color: var(--white-color);
  background-color: var(--gray-color);
  text-decoration: none;
}

.c1 {
  width: 100%;
  display: flex;
  flex-direction: column;
  align-items: center;
  text-align: center;
  row-gap: 16px;
}

.c2 {
  width: 100%;
  max-width: 400px;
}

.c3 {
  font-size: 36px;
  font-weight: 700;
}

.c0 {
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
  gap: 16px;
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <img
      alt="에러 발생"
      class="c2"
      src="/src/assets/images/empty.png"
    />
    <h2
      class="c3"
    >
      현재 등록된 기록이 없습니다.
    </h2>
    <button
      class="c4"
    >
      타이틀 화면으로 돌아가기
    </button>
    <button
      class="c4"
    >
      소통창구 (카카오 오픈채팅방)
    </button>
  </div>
</div>
`;

exports[`RecordSelection > should match snapshot with non-empty recordIds 1`] = `
.c4 {
  position: relative;
  width: 300px;
  min-height: 240px;
  display: grid;
  grid-template-rows: 200px 1fr;
  background-color: var(--primary-color);
  border-radius: 16px;
  overflow: hidden;
}

.c4:hover {
  background-color: var(--hovering-color);
}

.c4:hover>div>img {
  transform: scale(1.3);
}

.c4:hover>div>img+div {
  opacity: 1;
}

.c5 {
  width: 100%;
  height: 100%;
  overflow: hidden;
  border-top-left-radius: 16px;
  border-top-right-radius: 16px;
}

.c6 {
  width: 100%;
  height: 100%;
  object-fit: cover;
  transition: transform 100ms ease-in-out;
}

.c7 {
  width: 100%;
  height: 200px;
  position: absolute;
  left: 0;
  top: 0;
  background-color: rgba(255, 255, 255, 0.3);
  opacity: 0;
  transition: opacity 100ms ease-in-out;
}

.c8 {
  height: 100%;
  display: flex;
  flex-direction: column;
  row-gap: 8px;
  padding: 8px 16px;
  justify-content: center;
  align-items: flex-start;
}

.c9 {
  color: var(--white-color);
  font-size: 16px;
  font-weight: 600;
}

.c10 {
  color: var(--white-color);
  font-size: 16px;
  font-weight: 400;
}

.c0 {
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
  gap: 16px;
}

.c1 {
  font-size: 16px;
  font-weight: 400;
}

.c2 {
  width: 100%;
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
  justify-content: flex-start;
  align-items: flex-start;
  gap: 16px;
}

.c3 {
  color: unset;
  text-decoration: none;
}

<div
  class="c0"
>
  <span
    class="c1"
  >
    총 
    3
    건의 기록이 있습니다.
  </span>
  <div
    class="c2"
  >
    <a
      class="c3"
      href="/records/dodonpachi-cshot/2023-10-28"
    >
      <div
        class="c4"
      >
        <div
          class="c5"
        >
          <img
            alt="2023년 10월 28일"
            class="c6"
            src="Thumbnail URL"
          />
          <div
            class="c7"
          />
        </div>
        <div
          class="c8"
        >
          <span
            class="c9"
          >
            도돈파치 (1997): C-Shot
          </span>
          <span
            class="c9"
          >
            2023년 10월 28일
          </span>
          <span
            class="c10"
          >
            2-4 / 80000000점
          </span>
        </div>
      </div>
    </a>
    <a
      class="c3"
      href="/records/dodonpachi-cshot/2023-05-14"
    >
      <div
        class="c4"
      >
        <div
          class="c5"
        >
          <img
            alt="2023년 5월 14일"
            class="c6"
            src="Thumbnail URL"
          />
          <div
            class="c7"
          />
        </div>
        <div
          class="c8"
        >
          <span
            class="c9"
          >
            도돈파치 (1997): C-Shot
          </span>
          <span
            class="c9"
          >
            2023년 5월 14일
          </span>
          <span
            class="c10"
          >
            2-4 / 80000000점
          </span>
        </div>
      </div>
    </a>
    <a
      class="c3"
      href="/records/dodonpachi-cshot/2022-11-21"
    >
      <div
        class="c4"
      >
        <div
          class="c5"
        >
          <img
            alt="2022년 11월 21일"
            class="c6"
            src="Thumbnail URL"
          />
          <div
            class="c7"
          />
        </div>
        <div
          class="c8"
        >
          <span
            class="c9"
          >
            도돈파치 (1997): C-Shot
          </span>
          <span
            class="c9"
          >
            2022년 11월 21일
          </span>
          <span
            class="c10"
          >
            2-4 / 80000000점
          </span>
        </div>
      </div>
    </a>
  </div>
</div>
`;
